src/client/datascience/notebook/notebookControllerManager.ts

```diff
@@ -173,35 +183,51 @@ export class NotebookControllerManager implements INotebookControllerManager, IE
             return;
         }
 
-        // Prep so that we can track the selected controller for this document
-        this.controllerMapping.set(document, undefined);
-
         // Keep track of a token per document so that we can cancel the search if the doc is closed
         const preferredSearchToken = new CancellationTokenSource();
         this.findPreferredInProgress.set(document, preferredSearchToken);
 
-        this.findPreferredKernel(document, preferredSearchToken.token)
-            .then(async (preferredConnection) => {
-                if (preferredSearchToken.token.isCancellationRequested) {
-                    traceInfo('Find preferred kernel cancelled');
-                    return;
-                }
-                // await this.createOrDeletePlaceholderPythonKernel();
-
-                // If we found a preferred kernel, set the association on the NotebookController
-                if (preferredConnection) {
-                    traceInfo(
-                        `PreferredConnection: ${
-                            preferredConnection.id
-                        } found for NotebookDocument: ${document.uri.toString()}`
-                    );
-                    this.setPreferredController(document, preferredConnection).catch(traceError);
-                }
-            })
-            .finally(() => {
+        if (!this.isLocalLaunch) {
+            // For a remote connection check for new live kernel models before we find preferred
+            this.updateRemoteConnections(preferredSearchToken.token)
+                .then(() => {
+                    this.findPreferredController(document, preferredSearchToken.token).catch((error) => {
+                        traceError(error);
+                    });
+                })
+                .finally(() => {
+                    // Make sure that we clear our finding in progress when done
+                    this.findPreferredInProgress.delete(document);
+                });
+        } else {
+            this.findPreferredController(document, preferredSearchToken.token).finally(() => {
                 // Make sure that we clear our finding in progress when done
                 this.findPreferredInProgress.delete(document);
             });
+        }
+    }
+
+    // Find the preferred controller for the given notebook document
+    private async findPreferredController(document: NotebookDocument, cancelToken: CancellationToken) {
```

Comment:
Name feels off to me. It's setting the controller mapping, right? That's the goal of this function? Maybe setController?